README.md
# Cyanide

Cyanide is a BSON library for Elixir.

## Installation

The package can be installed by adding `cyanide` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:cyanide, "~> 1.0"}
  ]
end
```

## Usage

```elixir
Cyanide.encode(%{"value" => 42})
#=> {:ok, <<16, 0, 0, 0, 16, 118, 97, 108, 117, 101, 0, 42, 0, 0, 0, 0>>}

Cyanide.decode(<<16, 0, 0, 0, 16, 118, 97, 108, 117, 101, 0, 42, 0, 0, 0, 0>>)
#=> {:ok, %{"value" => 42}}
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.0.0] - 2023-06-20
### Added
- **BREAKING CHANGE**: new type for BSON binaries: `Cyanide.Binary`.
- Add `Cyanide.Encoder` protocol.

### Fixed
- Fixes from 1.0.1 (handling of malformed subdocuments).

## [1.0.1] - 2021-12-16
### Fixed
- Fix handling of malformed subdocuments.

## [1.0.0] - 2020-03-18
### Added
- First Cyanide release.


test/bson_test.exs
Code.require_file "test_helper.exs", __DIR__

defmodule Bson.Test do
  use ExUnit.Case

  doctest Bson
  doctest Bson.ObjectId
  doctest Bson.UTC
  doctest Bson.Decoder
  doctest Cyanide.Encoder.Float
  doctest Cyanide.Encoder.Integer
  doctest Cyanide.Encoder.Atom
  doctest Cyanide.Encoder.Bson.Regex
  doctest Cyanide.Encoder.Bson.ObjectId
  doctest Cyanide.Encoder.Bson.JS
  doctest Cyanide.Encoder.Bson.Bin
  doctest Cyanide.Encoder.Bson.Timestamp
  doctest Cyanide.Encoder.BitString
  doctest Cyanide.Encoder.Bson.UTC
  doctest Cyanide.Encoder.List
  doctest Cyanide.Encoder.Map

end


test/cyanide_test.exs
#
# This file is part of Cyanide.
#
# Copyright 2019 Ispirata Srl
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

defmodule CyanideTest do
  use ExUnit.Case

  test "encodes and decodes a map" do
    map1 = %{
      "t" => DateTime.utc_now() |> DateTime.to_unix(:millisecond),
      "v" => %{
        "some_binary" => %Cyanide.Binary{subtype: :generic, data: <<0, 0, 0, 0>>},
        "array" => [1, 2, 3, 4, 5],
        "some_bool" => true,
        "a_string" => "This is a string",
        "int_value0" => 1,
        "int_value1" => -1,
        "int_value2" => 1_152_921_504_606_846_976,
        "int_value3" => -1_152_921_504_606_846_976
      }
    }

    assert Cyanide.decode!(Cyanide.encode!(map1)) == map1
  end

  test "encodes and decodes a complex map" do
    map2 = %{
      "t" => DateTime.utc_now() |> DateTime.to_unix(:millisecond),
      "bin1" => %Cyanide.Binary{subtype: :generic, data: <<0, 1, 2, 3>>},
      "bin2" => %Cyanide.Binary{subtype: :generic, data: <<0>>},
      "bin3" => %Cyanide.Binary{subtype: :generic, data: <<>>},
      "map1" => %{},
      "map2" => %{"a" => true, "b" => false, "c" => nil},
      "map3" => %{
        "v" => [
          0,
          -1,
          1,
          2_147_483_647,
          2_147_483_648,
          -2_147_483_648,
          -2_147_483_649,
          9_223_372_036_854_775_807,
          -9_223_372_036_854_775_808
        ],
        "map4" => %{
          "map5" => %{
            "map6" => %{
              "v" => []
            }
          }
        }
      }
    }

    assert Cyanide.decode!(Cyanide.encode!(map2)) == map2
  end

  test "encodes and decodes a map made of special float values" do
    map3 = %{
      "v1" => :NaN,
      "v2" => :"-inf",
      "v3" => :inf
    }

    assert Cyanide.decode!(Cyanide.encode!(map3)) == map3
  end

  test "encodes and decodes a map made of floating point values" do
    map4 = %{
      "0" => 0,
      "a" => 1.1,
      "b" => 1.2,
      "c" => 1.3,
      "d" => 2,
      "e" => %{
        "some_value" => 5.2,
        "NaN" => :NaN,
        "inf" => :inf,
        "-inf" => :"-inf"
      },
      "f" => %{
        "some_value" => 6
      }
    }

    assert Cyanide.decode!(Cyanide.encode!(map4)) == map4
  end

  test "encodes and decodes an array of special float values" do
    map5 = %{
      "v" => [:NaN, :"-inf", :inf, :NaN]
    }

    assert Cyanide.decode!(Cyanide.encode!(map5)) == map5
  end

  test "encoding and decoding of an ascending ordered array" do
    o1 = %{
      "v" => [0, 1, 2, 3, 4]
    }

    assert Cyanide.decode!(Cyanide.encode!(o1)) == o1
  end

  test "encoding and decoding of a descending ordered array" do
    o2 = %{
      "v" => [5, 4, 3, 2, 1, 0]
    }

    assert Cyanide.decode!(Cyanide.encode!(o2)) == o2
  end

  test "encoding and decoding of an unordered array" do
    o3 = %{
      "v" => [0, 4, 3, 7, 1, 0]
    }

    assert Cyanide.decode!(Cyanide.encode!(o3)) == o3
  end
end


test/support/test1.ex
#
# This file is part of Cyanide.
#
# Copyright 2020 Ispirata Srl
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

defmodule Test1 do
  defstruct [
    :a,
    :b
  ]
end

defimpl Cyanide.Encoder, for: Test1 do
  def encode(value) do
    data =
      if value.a == value.b do
        1
      else
        0
      end

    %Cyanide.EncodedValue{
      tag: 0x08,
      data: <<data::8>>
    }
  end
end


mix.exs
#
# This file is part of Cyanide.
#
# Copyright 2019-2023 SECO Mind Srl
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

defmodule Cyanide.MixProject do
  use Mix.Project

  def project do
    [
      app: :cyanide,
      version: "2.0.0",
      elixir: "~> 1.6",
      elixirc_paths: elixirc_paths(Mix.env()),
      deps: deps(),
      package: package(),
      source_url: "https://github.com/secomind/cyanide",
      test_coverage: [tool: ExCoveralls],
      preferred_cli_env: [
        coveralls: :test,
        "coveralls.detail": :test,
        "coveralls.post": :test,
        "coveralls.html": :test
      ]
    ]
  end

  defp elixirc_paths(:test), do: ["lib", "test/support"]
  defp elixirc_paths(_), do: ["lib"]

  def application do
    [
      extra_applications: [:logger]
    ]
  end

  defp deps do
    [
      {:excoveralls, "~> 0.9.1", only: :test},
      {:dialyxir, "== 0.5.1", only: [:dev], runtime: false},
      {:ex_doc, ">= 0.0.0", only: :dev, runtime: false}
    ]
  end

  defp package do
    [
      description: "An Elixir BSON encoding/decoding library.",
      maintainers: ["Davide Bettio", "Riccardo Binetti"],
      licenses: ["Apache-2.0"],
      links: %{
        "GitHub" => "https://github.com/secomind/cyanide",
        "Documentation" => "http://hexdocs.pm/cyanide/"
      }
    ]
  end
end


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure your application as:
#
#     config :cyanide, key: :value
#
# and access this configuration in your application as:
#
#     Application.get_env(:cyanide, :key)
#
# You can also configure a 3rd-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env()}.exs"


lib/cyanide/binary.ex
#
# This file is part of Cyanide.
#
# Copyright 2020 Ispirata Srl
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

defmodule Cyanide.Binary do
  @moduledoc """
  Represents a binary with a subtype.

  ## Fields
  * `subtype` is any of `:generic`, `:function`, `:old_binary`, `:old_uuid`, `:uuid`, `:md5`,
  `:encrypted_bson`, or any user defined value (any integer >= `0x80` and <= `0xFF`).
  * `data` is a binary (such as `<<0, 1, 2>>` or `"test"`).
  """

  defstruct [
    :subtype,
    :data
  ]

  @type subtype() ::
          :generic
          | :function
          | :old_binary
          | :old_uuid
          | :uuid
          | :md5
          | :encrypted_bson
          | pos_integer()

  @type t() :: %__MODULE__{
          subtype: subtype(),
          data: binary()
        }

  @spec cast_subtype(byte) :: {:ok, subtype()} | :error
  def cast_subtype(value) when is_integer(value) and value < 0xFF do
    case value do
      0 -> {:ok, :generic}
      1 -> {:ok, :function}
      2 -> {:ok, :old_binary}
      3 -> {:ok, :old_uuid}
      4 -> {:ok, :uuid}
      5 -> {:ok, :md5}
      6 -> {:ok, :encrypted_bson}
      x when x >= 0x80 -> {:ok, x}
      _ -> :error
    end
  end
end


lib/cyanide/encoder.ex
#
# This file is part of Cyanide.
#
# Copyright 2020 Ispirata Srl
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

defprotocol Cyanide.Encoder do
  @moduledoc """
  Protocol controlling how a value is encoded to BSON.

  An encode implementation can either return any encodable value or a Cyanide.EncodedValue struct.
  In most of the cases returning a map is enough, Cyanide.EncodedValue is meant to be used when a
  specific BSON type should be enforced.
  Only maps are supported for root document encoding.

  ## Example

  Let's assume a presence of the following struct:

      defmodule Test do
        defstruct [:value1, :value2]
      end

  It can be encoded using the following implementation:

      defimpl Cyanide.Encoder, for: Test do
        def encode(value) do
          Map.to_struct(value)
        end
      end

  Let's assume that a small value should be always encoded as Int64.
  This problem can be solved by implementing a wrapper struct.

      defmodule Int64 do
        defstruct [:value]
      end

  It can be encoded using the following implementation:

      defimpl Cyanide.Encoder, for: Int64 do
        def encode(wrapped) do
          value = wrapped.value

          %Cyanide.EncodedValue{
            tag: 0x12,
            data: <<value::signed-little-64>>
          }
        end
      end
  """

  @spec encode(term()) :: Cyanide.bson_type() | Cyanide.EncodedValue.t()
  def encode(value)
end


lib/cyanide/encoded_value.ex
#
# This file is part of Cyanide.
#
# Copyright 2020 Ispirata Srl
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

defmodule Cyanide.EncodedValue do
  @moduledoc """
  Wraps a BSON encoded value.

  ## Fields
  * `tag` is a type integer tag (such as 0x01, 0x12, ...) as described on the BSON specification.
  * `data` is the raw encoded value iodata.
  """

  defstruct [
    :tag,
    :data
  ]

  @type t() :: %__MODULE__{
          tag: integer(),
          data: iodata()
        }
end


lib/bson_encoder.ex
defprotocol Cyanide.Encoder do
  @moduledoc """
  `Cyanide.Encoder` protocol defines Bson encoding according to Elxir terms and some Bson predefined structs (see `Bson`).

  List of the protocol implementations:

  * `Map` - Encodes a map into a document
  * `HasDict` - Encodes a HashDict into a document
  * `Keyword` - Encodes a Keyword into a document
  *  `List` - Encodes a list of key-alue pairs into a document otherwize encode list into array
  * `Integer` - Encodes integer in 32 or 64 bits
  * `Float` - Encodes float in 64 bits
  * `Atom` - Encodes special atom (`false`, `true`, `nil`,
  `:nan`, `:+inf`, `:-inf`, `MIN_KEY` and `MAX_KEY`) in appropriate format
  others in special type Symbol
  * `BitString` - as binary string
  * `Bson.Regex' - see specs
  * `Bson.ObjectId' - see specs
  * `Bson.JS' - see specs
  * `Bson.Bin' - see specs
  * `Bson.Timestamp  ' - see specs
  """

  @doc """
  Returns a binary representing a term in Bson format
  """
  def encode(term)
end

defmodule Bson.Encoder do
  defmodule Error do
    @moduledoc """
    Container for error messages

    * `what` has triggerred the error
    * `acc` contains what was already decoded for this term (ie the size of a string when the string itself could not be decoded)
    * `term` that failed to be encoded
    """
    defstruct [what: nil, acc: [], term: nil]
    defimpl Inspect, for: Error do
      def inspect(e,_), do: inspect([what: e.what, term: e.term, acc: e.acc])
    end
  end

  @doc """
  Creates a document using a collection of element, this is, a key-value pair
  """
  def document(element_list) do
    case Enumerable.reduce(element_list, {:cont, []},
      fn({key, value}, acc) when is_binary(key) -> accumulate_elist(key, value, acc)
        ({key, value}, acc) when is_atom(key) -> accumulate_elist(Atom.to_string(key), value, acc)
        (element, acc) -> {:halt, %Error{what: [:element], term: element, acc: acc |> Enum.reverse}}
      end) do
      {:halted, error} -> error
      {:done, acc} ->
        acc |> Enum.reverse |> IO.iodata_to_binary |> wrap_document
    end
  end

  defimpl Cyanide.Encoder, for: Integer do
    @doc """
    iex> Cyanide.Encoder.encode(2)
    {<<16>>, <<2, 0, 0, 0>>}
    iex> Cyanide.Encoder.encode(-2)
    {<<16>>, <<254, 255, 255, 255>>}
    iex> Cyanide.Encoder.encode -0x80000001
    {<<18>>, <<255, 255, 255, 127, 255, 255, 255, 255>>}

    iex> Cyanide.Encoder.encode 0x8000000000000001
    %Bson.Encoder.Error{what: [Integer], term: 0x8000000000000001}
    """
    def encode(i) when i >= -2_147_483_648 and i <= 2_147_483_647, do: {<<0x10>>, <<i::32-signed-little>>}
    def encode(i) when i >= -9_223_372_036_854_775_808 and i <= 9_223_372_036_854_775_807, do: {<<0x12>>, <<i::64-signed-little>>}
    def encode(i), do: %Error{what: [Integer], term: i}
  end

  defimpl Cyanide.Encoder, for: Float do
    @doc """
    iex> Cyanide.Encoder.encode(1.1)
    {<<1>>, <<154, 153, 153, 153, 153, 153, 241, 63>>}
    """
    def encode(f), do: {<<0x01>>, <<(f)::size(64)-float-little>>}
  end

  defimpl Cyanide.Encoder, for: Atom do
    @doc """
    iex> Cyanide.Encoder.encode(true)
    {<<8>>, <<1>>}
    iex> Cyanide.Encoder.encode(nil)
    {<<10>>, <<>>}
    iex> Cyanide.Encoder.encode(:max_key)
    {<<127>>, <<>>}
    iex> Cyanide.Encoder.encode(:min_key)
    {<<255>>, <<>>}
    iex> Cyanide.Encoder.encode(:nan)
    {<<1>>, <<0, 0, 0, 0, 0, 0, 248, 127>>}
    iex> Cyanide.Encoder.encode(:'+inf')
    {<<1>>, <<0, 0, 0, 0, 0, 0, 240, 127>>}
    iex> Cyanide.Encoder.encode(:'-inf')
    {<<1>>, <<0, 0, 0, 0, 0, 0, 240, 255>>}
    iex> Cyanide.Encoder.encode(:atom)
    {<<14>>, [<<5, 0, 0, 0>>, "atom", <<0>>]}

    """
    # predefind Bson value
    def encode(false),   do: {<<0x08>>, <<0x00>>}
    def encode(true),    do: {<<0x08>>, <<0x01>>}
    def encode(nil),     do: {<<0x0a>>, <<>>}
    def encode(:nan),    do: {<<0x01>>, <<0, 0, 0, 0, 0, 0, 248, 127>>}
    def encode(:'+inf'), do: {<<0x01>>, <<0, 0, 0, 0, 0, 0, 240, 127>>}
    def encode(:'-inf'), do: {<<0x01>>, <<0, 0, 0, 0, 0, 0, 240, 255>>}
    def encode(:min_key), do: {<<0xff>>, <<>>}
    def encode(:max_key), do: {<<0x7f>>, <<>>}
    # other Elixir atom are encoded like strings ()
    def encode(atom),    do: {<<0x0e>>, (atom |> Atom.to_string |> Bson.Encoder.wrap_string)}
  end

  defimpl Cyanide.Encoder, for: Bson.UTC do
    @doc """
    iex> Cyanide.Encoder.encode(Bson.UTC.from_now({1390, 324703, 518471}))
    {<<9>>, <<30, 97, 207, 181, 67, 1, 0, 0>>}
    """
    def encode(%Bson.UTC{ms: ms}) when is_integer(ms), do: {<<0x09>>, <<ms::64-little-signed>>}
    def encode(utc), do: %Error{what: [Bson.UTC], term: utc}
  end

  defimpl Cyanide.Encoder, for: Bson.Regex do
    @doc """
    iex> Cyanide.Encoder.encode(%Bson.Regex{pattern: "p", opts: "i"})
    {<<11>>, ["p", <<0>>, "i", <<0>>]}
    """
    def encode(%Bson.Regex{pattern: p, opts: o}) when is_binary(p) and is_binary(o), do: {<<0x0b>>, [p, <<0x00>>, o, <<0x00>>]}
    def encode(regex), do: %Error{what: [Bson.Regex], term: regex}
  end

  defimpl Cyanide.Encoder, for: Bson.ObjectId do
    @doc """
    iex> Cyanide.Encoder.encode(%Bson.ObjectId{oid: <<0xFF>>})
    {<<0x07>>, <<255>>}

    iex> Cyanide.Encoder.encode(%Bson.ObjectId{oid: 123})
    %Bson.Encoder.Error{what: [Bson.ObjectId], term: %Bson.ObjectId{oid: 123}}
    """
    def encode(%Bson.ObjectId{oid: oid}) when is_binary(oid), do: {<<0x07>>, oid}
    def encode(oid), do: %Error{what: [Bson.ObjectId], term: oid}
  end

  defimpl Cyanide.Encoder, for: Bson.JS do
    @doc """
    iex> Cyanide.Encoder.encode(%Bson.JS{code: "1+1;"})
    {<<13>>, [<<5, 0, 0, 0>>, "1+1;", <<0>>]}
    iex> Cyanide.Encoder.encode(%Bson.JS{code: "1+1;", scope: %{a: 0, b: "c"}})
    {<<15>>, <<34, 0, 0, 0, 5, 0, 0, 0, 49, 43, 49, 59, 0, 21, 0, 0, 0, 16, 97, 0, 0, 0, 0, 0, 2, 98, 0, 2, 0, 0, 0, 99, 0, 0>>}
    """
    def encode(%Bson.JS{code: js, scope: nil}) when is_binary(js) do
      {<<0x0d>>, Bson.Encoder.wrap_string(js)}
    end
    def encode(%Bson.JS{code: js, scope: ctx}) when is_binary(js) and is_map(ctx) do
      case Bson.Encoder.document(ctx) do
        %Error{}=error -> %Error{error|what: {:js_context, error.what}}
        ctxBin ->
          {<<0x0f>>, [Bson.Encoder.wrap_string(js), ctxBin] |> IO.iodata_to_binary |> js_ctx}
      end
    end
    def encode(js), do: %Error{what: [Bson.JS], term: js}

    defp js_ctx(jsctx), do: <<(byte_size(jsctx)+4)::32-little-signed, jsctx::binary>>
  end

  defimpl Cyanide.Encoder, for: Bson.Bin do
    @doc """
    iex> Cyanide.Encoder.encode(%Bson.Bin{bin: "e", subtype: Bson.Bin.subtyx(:user)})
    {<<5>>,[<<1, 0, 0, 0>>, 128, "e"]}
    """
    def encode(%Bson.Bin{bin: bin, subtype: subtype}), do: encode(bin, subtype)
    def encode(bin, subtype)
      when is_binary(bin) and is_integer(subtype),
      do:  {<<0x05>>, [<<byte_size(bin)::32-little-signed>>, subtype, bin]}
    def encode(bin, subtype), do: %Error{what: [Bson.Bin], term: {bin, subtype}}
  end

  defimpl Cyanide.Encoder, for: Bson.Timestamp do
    @doc """
    iex> Cyanide.Encoder.encode(%Bson.Timestamp{inc: 1, ts: 2})
    {<<17>>,<<1, 0, 0, 0, 2, 0, 0, 0>>}
    """
    def encode(%Bson.Timestamp{inc: i, ts: t})
      when is_integer(i) and -0x80000000 <= i and i <= 0x80000000
       and is_integer(t) and -0x80000000 <= t and t <= 0x80000000,
      do: {<<0x11>>, <<i::32-signed-little, t::32-signed-little>>}
      def encode(ts), do: %Error{what: [Bson.Timestamp], term: ts}
  end

  defimpl Cyanide.Encoder, for: BitString do
    @doc """
    iex> Cyanide.Encoder.encode("a")
    {<<2>>, [<<2, 0, 0, 0>>, "a", <<0>>]}
    """
    def encode(s) when is_binary(s),  do: {<<0x02>>, Bson.Encoder.wrap_string(s)}
    def encode(bits), do: %Error{what: [BitString], term: bits}
  end

  defimpl Cyanide.Encoder, for: List do
    @doc """
    iex> Cyanide.Encoder.encode([])
    {<<4>>,<<5, 0, 0, 0, 0>>}
    iex> Cyanide.Encoder.encode([2, 3])
    {<<4>>,<<19, 0, 0, 0, 16, 48, 0, 2, 0, 0, 0, 16, 49, 0, 3, 0, 0, 0, 0>>}
    iex> Cyanide.Encoder.encode([1,[nil]])
    {<<4>>,<<23, 0, 0, 0, 16, 48, 0, 1, 0, 0, 0, 4, 49, 0, 8, 0, 0, 0, 10, 48, 0, 0, 0>>}
    iex> Cyanide.Encoder.encode([1,[2, 3]])
    {<<4>>,<<34, 0, 0, 0, 16, 48, 0, 1, 0, 0, 0, 4, 49, 0, 19, 0, 0, 0, 16, 48, 0, 2, 0, 0, 0, 16, 49, 0, 3, 0, 0, 0, 0, 0>>}

    # Keyword and list of key-value pairs
    iex> Cyanide.Encoder.encode([a: "r"])
    {<<3>>,<<14, 0, 0, 0, 2, 97, 0, 2, 0, 0, 0, 114, 0, 0>>}
    iex> Cyanide.Encoder.encode([{"a", "s"}])
    {<<3>>,<<14, 0, 0, 0, 2, 97, 0, 2, 0, 0, 0, 115, 0, 0>>}

    iex> Cyanide.Encoder.encode([{"a", "s"}, {:b, "r"}, 1, 2])
    %Bson.Encoder.Error{
      term: 1,
      what: [:element],
      acc: [[<<2>>, "a", <<0>>, [<<2, 0, 0, 0>>, "s", <<0>>]],
            [<<2>>, "b", <<0>>, [<<2, 0, 0, 0>>, "r", <<0>>]]]}

    iex> Cyanide.Encoder.encode([2, 3, ])
    {<<4>>,<<19, 0, 0, 0, 16, 48, 0, 2, 0, 0, 0, 16, 49, 0, 3, 0, 0, 0, 0>>}

    """
    def encode([{k, _}|_]=elist) when is_atom(k) or is_binary(k) do
      case Bson.Encoder.document(elist) do
        %Error{}=error -> error
        encoded_elist -> {<<0x03>>, encoded_elist}
      end
    end
    def encode(list) do
      case Bson.Encoder.array(list) do
        %Error{}=error -> error
        encoded_list -> {<<0x04>>, encoded_list}
      end
    end
  end

  defimpl Cyanide.Encoder, for: [Map, HashDict, Keyword] do
    @doc """
    # Map
    iex> Cyanide.Encoder.encode(%{})
    {<<3>>,<<5, 0, 0, 0, 0>>}
    iex> Cyanide.Encoder.encode(%{a: "r"})
    {<<3>>,<<14, 0, 0, 0, 2, 97, 0, 2, 0, 0, 0, 114, 0, 0>>}
    iex> Cyanide.Encoder.encode(%{a: 1, b: 5})
    {<<3>>,<<19, 0, 0, 0, 16, 97, 0, 1, 0, 0, 0, 16, 98, 0, 5, 0, 0, 0, 0>>}
    iex> Cyanide.Encoder.encode(%{a: 1, b: %{c: 3}})
    {<<3>>,<<27, 0, 0, 0, 16, 97, 0, 1, 0, 0, 0, 3, 98, 0, 12, 0, 0, 0, 16, 99, 0, 3, 0, 0, 0, 0, 0>>}

    # HashDict
    iex> Cyanide.Encoder.encode(%HashDict{})
    {<<3>>,<<5, 0, 0, 0, 0>>}
    iex> Cyanide.Encoder.encode(HashDict.put(%HashDict{}, :a, "r"))
    {<<3>>,<<14, 0, 0, 0, 2, 97, 0, 2, 0, 0, 0, 114, 0, 0>>}

    iex> Cyanide.Encoder.encode(%{a: "va", b: "vb", u: %Bson.UTC{ms: "e"}})
    %Bson.Encoder.Error{
      what: ["u", Bson.UTC],
      term: %Bson.UTC{ms: "e"},
      acc: [[[<<2>>, "a", <<0>>, [<<3, 0, 0, 0>>, "va", <<0>>]],
             [<<2>>, "b", <<0>>, [<<3, 0, 0, 0>>, "vb", <<0>>]]]]}
    iex> Cyanide.Encoder.encode([1, 2, %Bson.UTC{ms: "e"}])
    %Bson.Encoder.Error{
      what: ["2", Bson.UTC],
      term: %Bson.UTC{ms: "e"},
      acc: [[[<<16>>, "0", <<0>>, <<1, 0, 0, 0>>],
             [<<16>>, "1", <<0>>, <<2, 0, 0, 0>>]]]}
    iex> Cyanide.Encoder.encode(%{a: "va", b: "vb", c: %{c1: "vc1", cu: %Bson.UTC{ms: "e"}}})
    %Bson.Encoder.Error{
      what: ["c", "cu", Bson.UTC],
      term: %Bson.UTC{ms: "e"},
      acc: [[[<<2>>, "a", <<0>>, [<<3, 0, 0, 0>>, "va", <<0>>]],
            [<<2>>, "b", <<0>>,  [<<3, 0, 0, 0>>, "vb", <<0>>]]],
           [[<<2>>, "c1", <<0>>, [<<4, 0, 0, 0>>, "vc1", <<0>>]]]]}
    iex> Cyanide.Encoder.encode(%{a: "va", b: "vb", c: ["c0", %Bson.UTC{ms: "e"}]})
    %Bson.Encoder.Error{
      what: ["c", "1", Bson.UTC],
      term: %Bson.UTC{ms: "e"},
      acc: [[[<<2>>, "a", <<0>>, [<<3, 0, 0, 0>>, "va", <<0>>]],
            [<<2>>, "b", <<0>>, [<<3, 0, 0, 0>>, "vb", <<0>>]]],
           [[<<2>>, "0", <<0>>, [<<3, 0, 0, 0>>, "c0", <<0>>]]]]}

    """
    def encode(dict) do
      case Bson.Encoder.document(dict) do
        %Error{}=error -> error
        encoded_dict -> {<<0x03>>, encoded_dict}
      end
    end
  end

  @doc """
  Creates a document for an array (list of items)
  """
  def array(item_list) do
    case Enumerable.reduce(item_list, {:cont, {[], 0}},
      fn(item, {acc, i}) ->
        case accumulate_elist(Integer.to_string(i), item, acc) do
          {:cont, acc} -> {:cont, {acc, i+1}}
          {:halt, error} -> {:halt, error}
        end
      end) do
      {:halted, error} -> error
      {:done, {bufferAcc, _}} ->
        bufferAcc |> Enum.reverse |> IO.iodata_to_binary |> wrap_document
    end
  end

  @doc """
  Wraps a bson document with size and trailing null character
  """
  def wrap_document(elist), do: <<(byte_size(elist)+5)::32-little-signed>> <> elist <> <<0x00>>

  @doc """
  Wraps a bson document with size and trailing null character
  """
  def wrap_string(string), do: [<<(byte_size(string)+1)::32-little-signed>>, string, <<0x00>>]

  @doc """
  Accumulate element in an element list
  """
  def accumulate_elist(name, value, elist) do
    case element(name, value) do
      %Error{}=error -> {:halt, %Error{error|acc: [Enum.reverse(elist)|error.acc]}}
      encoded_element -> {:cont, [encoded_element | elist]}
    end
  end

  @doc """
  Returns encoded element using its name and value
  """
  def element(name, value) do
    case Cyanide.Encoder.encode(value) do
      %Error{}=error -> %Error{error|what: [name|error.what]}
      {kind, encoded_value} -> [kind, name, <<0x00>>, encoded_value]
    end
  end
end
